components/viz/service/display/starboard/overlay_strategy_underlay_starboard.cc

```diff
@@ -0,0 +1,196 @@
+// Copyright 2024 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "components/viz/service/display/starboard/overlay_strategy_underlay_starboard.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/containers/adapters.h"
+#include "base/logging.h"
+#include "components/viz/common/quads/draw_quad.h"
+#include "components/viz/common/quads/solid_color_draw_quad.h"
+#include "components/viz/common/quads/video_hole_draw_quad.h"
+#include "components/viz/service/display/overlay_candidate_factory.h"
+#include "ui/gfx/geometry/rect_conversions.h"
+
+namespace viz {
+
+OverlayStrategyUnderlayStarboard::OverlayStrategyUnderlayStarboard(
+    OverlayProcessorUsingStrategy* capability_checker)
+    : OverlayStrategyUnderlay(capability_checker) {}
+
+OverlayStrategyUnderlayStarboard::~OverlayStrategyUnderlayStarboard() {}
+
+void OverlayStrategyUnderlayStarboard::Propose(
+    const SkM44& output_color_matrix,
+    const OverlayProcessorInterface::FilterOperationsMap& render_pass_filters,
+    const OverlayProcessorInterface::FilterOperationsMap&
+        render_pass_backdrop_filters,
+    DisplayResourceProvider* resource_provider,
+    AggregatedRenderPassList* render_pass_list,
+    SurfaceDamageRectList* surface_damage_rect_list,
+    const PrimaryPlane* primary_plane,
+    std::vector<OverlayProposedCandidate>* candidates,
+    std::vector<gfx::Rect>* content_bounds) {
+  auto* render_pass = render_pass_list->back().get();
+  QuadList& quad_list = render_pass->quad_list;
+  OverlayCandidate candidate;
+  auto overlay_iter = quad_list.end();
+  OverlayCandidateFactory candidate_factory = OverlayCandidateFactory(
+      render_pass, resource_provider, surface_damage_rect_list,
+      &output_color_matrix, GetPrimaryPlaneDisplayRect(primary_plane),
+      &render_pass_filters);
+
+  // Original code did reverse iteration.
+  // Here we do forward but find the last one. which should be the same thing.
+  for (auto it = quad_list.begin(); it != quad_list.end(); ++it) {
+    if (OverlayCandidate::IsInvisibleQuad(*it)) {
+      continue;
+    }
+
+    // Look for quads that are overlayable and require an overlay. Chromecast
```

Comment:
Let's create a separate PR, copy the original file into this file without any modification, and without including them in build, so we can review the changes and potential keep some history.

Reply:
Done.